### PR TITLE
analyze: recover from failures in shim generation

### PR DESCRIPTION
**Pull request by spernsteiner**
Previously, if `bad` would call `good` through a shim, but generation of `good_shim` fails (for example, because the shim would require an unsupported cast like `*mut T -> &[T]`), then `bad` would still be rewritten to call the nonexistent `good_shim` function.  With this branch, we instead mark `good` as failed so that `good_shim` is no longer needed (shims are only needed for calling non-failed functions) and retry the entire rewriting pass.  Marking a function as failed may require shims to be generated for its callees, but eventually this process will converge once all the functions where shims can't be generated have been marked as failed.

### CODE REVIEW

File: c2rust-analyze/src/main.rs
Comment (aneksteind) [inline c2rust-analyze/src/main.rs:814]: nit: I think something like `START REWRITE LOOP` would be better

File: c2rust-analyze/tests/filecheck/unrewritten_calls_shim_fail.rs
Comment (aneksteind) [inline c2rust-analyze/tests/filecheck/unrewritten_calls_shim_fail.rs:6]: should this assert that a cast is present to call `slice_bad`?

**Review by aneksteind — APPROVED** [review]
LGTM, but if the test could use an assertion for the cast then it'd be good to put that before merging

File: c2rust-analyze/tests/filecheck/unrewritten_calls_shim_fail.rs
Comment (spernsteiner) [reply c2rust-analyze/tests/filecheck/unrewritten_calls_shim_fail.rs:6]: Cast insertion is already checked by `unrewritten_calls.rs`; this test is only checking the retry behavior when shim generation fails.

File: c2rust-analyze/src/main.rs
Comment (spernsteiner) [reply c2rust-analyze/src/main.rs]: I cleaned up the debug prints a little bit

**Pull request merged**
